ros2action/ros2action/api/__init__.py

```diff
@@ -0,0 +1,147 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from ament_index_python import get_resource
+from ament_index_python import get_resources
+from ament_index_python import has_resource
+
+import rclpy.action
+from rclpy.expand_topic_name import expand_topic_name
+from rclpy.validate_full_topic_name import validate_full_topic_name
+from ros2cli.node.direct import DirectNode
+
+
+def _is_action_status_topic(topic_name, action_name):
+    return action_name + '/_action/status' == topic_name
+
+
+def get_action_clients_and_servers(*, node, action_name):
+    action_clients = []
+    action_servers = []
+
+    try:
+        expanded_name = expand_topic_name(action_name, node.get_name(), node.get_namespace())
+    except ValueError as e:
+        raise RuntimeError(e)
+    try:
+        validate_full_topic_name(expanded_name)
+    except rclpy.exceptions.InvalidTopicNameException as e:
+        raise RuntimeError(e)
+
+    node_names_and_ns = node.get_node_names_and_namespaces()
+    for node_name, node_ns in node_names_and_ns:
+        # Construct fully qualified name
+        node_fqn = '/'.join(node_ns) + node_name
+
+        # Get any action clients associated with the node
+        client_names_and_types = rclpy.action.get_action_client_names_and_types_by_node(
+            node,
+            node_name,
+            node_ns,
+        )
+        for client_name, client_types in client_names_and_types:
+            if client_name == expanded_name:
+                action_clients.append((node_fqn, client_types))
+
+        # Get any action servers associated with the node
+        server_names_and_types = rclpy.action.get_action_server_names_and_types_by_node(
+            node,
+            node_name,
+            node_ns,
+        )
+        for server_name, server_types in server_names_and_types:
+            if server_name == expanded_name:
+                action_servers.append((node_fqn, server_types))
+
+    return (action_clients, action_servers)
+
+
+def get_action_names_and_types(*, node):
+    return rclpy.action.get_action_names_and_types(node)
+
+
+def get_action_names(*, node):
+    action_names_and_types = get_action_names_and_types(node=node)
+    return [n for (n, t) in action_names_and_types]
+
+
+def get_action_types(package_name):
+    if not has_resource('packages', package_name):
+        raise LookupError('Unknown package name')
+    try:
+        content, _ = get_resource('rosidl_interfaces', package_name)
+    except LookupError:
+        return []
+    interface_names = content.splitlines()
+    # TODO(jacobperron) this logic should come from a rosidl related package
+    # Only return actions in action folder
+    return list(sorted({
+        n[7:-7]
+        for n in interface_names
+        if n.startswith('action/') and n[-7:] in ('.idl', '.action')}))
```

Comment:
How about moving this into `rosidl_runtime_py`?

(Could be done in a follow up PR.)

Same for the similar commented blocks below.

Reply:
I think that makes sense. Created a ticket to track this https://github.com/ros2/ros2cli/issues/218